src/main/java/roomescape/dao/MemberDao.java

```diff
@@ -0,0 +1,75 @@
+package roomescape.dao;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.sql.DataSource;
+import org.springframework.dao.support.DataAccessUtils;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
+import org.springframework.stereotype.Repository;
+import roomescape.dao.rowmapper.MemberRowMapper;
+import roomescape.domain.Member;
+import roomescape.domain.MemberRepository;
+
+@Repository
+public class MemberDao implements MemberRepository {
+    private final JdbcTemplate jdbcTemplate;
+    private final SimpleJdbcInsert jdbcInsert;
+    private final MemberRowMapper memberRowMapper;
+
+    public MemberDao(JdbcTemplate jdbcTemplate, DataSource dataSource, MemberRowMapper memberRowMapper) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.jdbcInsert = new SimpleJdbcInsert(dataSource)
+                .withTableName("member")
+                .usingGeneratedKeyColumns("id");
+        this.memberRowMapper = memberRowMapper;
+    }
+
+    @Override
+    public List<Member> findAll() {
+        String sql = "SELECT * FROM member";
+        return jdbcTemplate.query(sql, memberRowMapper);
+    }
+
+    @Override
+    public Optional<Member> findById(Long id) {
+        String sql = "SELECT * FROM member WHERE id = ?";
+        List<Member> member = jdbcTemplate.query(sql, memberRowMapper, id);
+        return DataAccessUtils.optionalResult(member);
```

Comment:
처음에 해피케이스만 고려할 땐 아래와 같이 구현해 사용했는데, 후에 id에 해당하는 Reservation이 없는 경우를 처리하려 하니 아래 로직이 제대로 동작하지 않는 걸 확인할 수 있었습니다.

```java
@Override
public Optional<Reservation> findById(Long id) {
    Reservation reservation = jdbcTemplate.queryForObject(SQL, reservationRowMapper, id);
    return Optional.ofNullable(reservation);
}
```

queryForObject는 찾을 값이 존재하지 않으면  `EmptyResultDataAccessException` 을 던지기 때문입니다. 따라서 일단 이를 단순히 캐치해서 해결해보았습니다.

```java
@Override
public Optional<Reservation> findById(Long id) {
    try {
	return Optional.of(jdbcTemplate.queryForObject(SQL, reservationRowMapper, id));
    } catch (EmptyResultDataAccessException e) {
        return Optional.empty();
    }
}
```

영 찝찝해서 queryForObject가 `EmptyResultDataAccessException`을 뱉는 이유를 찾아보기로 했습니다. `queryForObject()`의 소스를 보니 내부적으로 `query()` 메서드를 써서 List 형태로 값을 찾는 걸 확인할 수 있었습니다.

```java
public <T> T queryForObject(String sql, RowMapper<T> rowMapper, @Nullable Object... args) throws DataAccessException {
	List<T> results = query(sql, args, new RowMapperResultSetExtractor<>(rowMapper, 1));
	return DataAccessUtils.nullableSingleResult(results);
}
```

그리고 이 List를 처리하는 `DataAccessUtils.nullableSingleResult()`의 소스를 보니 List가 비어있을 때 `EmptyResultDataAccessException` 를 뱉고 있는 걸 확인할 수 있었습니다.

```java
public static <T> T nullableSingleResult(@Nullable Collection<T> results) throws IncorrectResultSizeDataAccessException {
	// This is identical to the requiredSingleResult implementation but differs in the
	// semantics of the incoming Collection (which we currently can't formally express)
	if (CollectionUtils.isEmpty(results)) {
		throw new EmptyResultDataAccessException(1);
	}
	if (results.size() > 1) {
		throw new IncorrectResultSizeDataAccessException(1, results.size());
	}
	return results.iterator().next();
}
```

아마 jdbcTemplate 을 설계한 사람은 queryForobject는 반드시 하나의 결과가 있어야 한다고 생각한 거 같습니다. 그리고 이에 대한 논의가 실제로 광범위하게 이뤄졌다 합니다. 아래 issue 에서 그 논의에 대한 결론을 축약적으로 볼 수 있습니다.

https://github.com/spring-projects/spring-framework/issues/30927

결론적으로 저는 queryForObject 대신 query 를 직접 사용해 DataAccessUtils.nullableSingleResult 가 아닌 DataAccessUtils.optionalResult()를 사용해 문제를 해결해줬습니다.

> Granted, there is the stream closing topic. Even terminal operations do not close the stream, so there has to be a rather inconvenient `try` block against any shortcut terminal operations which makes it less attractive for chaining indeed. I suppose stylistically, that makes the `DataAccessUtils.optionalResult` helper the best present option (with no-more-than-1-row enforcement which we usually want), in combination with `JdbcTemplate.queryForList` since combining it with `queryForStream` does not provide additional value.
>

Reply:
💯 매우 좋아요.

코드를 작성하는 노력도 필요하지만 잘 읽고 이해하는 노력도 필요합니다. 라이브러리 코드를 읽을 줄 알고 적절히 활용할 수 있으면 도라는 자유로워질거에요.